cmd/lotus-shed/fip-0036.go

```diff
@@ -0,0 +1,565 @@
+		if err != nil {
+			return xerrors.Errorf("fail to get init actor: ", err)
+		}
+
+		initState, err := _init.Load(store, ia)
+		if err != nil {
+			return xerrors.Errorf("fail to load init state: ", err)
+		}
+
+		initMap, err := initState.AddressMap()
+		if err != nil {
+			return xerrors.Errorf("fail to load init map: ", err)
+		}
+
+		lookupId := func(addr address.Address) address.Address {
+			if addr.Protocol() == address.ID {
+				return addr
+			}
+
+			var actorID cbg.CborInt
+			if found, err := initMap.Get(abi.AddrKey(addr), &actorID); err != nil {
+				panic(err)
+			} else if found {
+				// Reconstruct address from the ActorID.
+				idAddr, err := address.NewIDAddress(uint64(actorID))
+				if err != nil {
+					panic(err)
+				}
+				return idAddr
+			}
+
+			panic("didn't find addr")
+		}
+
+		// we need to build several pieces of information, as we traverse the state tree:
+		// a map of accounts to every msig that they are a signer of
+		accountsToMultisigs := make(map[address.Address][]address.Address)
+		// a map of multisigs to some info about them for quick lookup
+		msigActorsInfo := make(map[address.Address]msigBriefInfo)
+		// a map of accounts to every miner that they are an owner of
+		ownerMap := make(map[address.Address][]address.Address)
+		// a map of accounts to every miner that they are a worker of
+		workerMap := make(map[address.Address][]address.Address)
+		// a map of miners to some info aboout them for quick lookup
+		minerActorsInfo := make(map[address.Address]minerBriefInfo)
+		// a map of client addresses to deal data stored in proposals
+		clientToDealStorage := make(map[address.Address]abi.StoragePower)
+
+		fmt.Println("iterating over all actors")
+		count := 0
+		err = st.ForEach(func(addr address.Address, act *types.Actor) error {
+			if count%200000 == 0 {
+				fmt.Println("processed ", count, " actors building maps")
+			}
+			count++
+			if builtin.IsMultisigActor(act.Code) {
+				ms, err := multisig.Load(store, act)
+				if err != nil {
+					return fmt.Errorf("load msig failed %v", err)
+
+				}
+
+				// TODO: Confirm that these are always ID addresses
+				signers, err := ms.Signers()
+				if err != nil {
+					return xerrors.Errorf("fail to get msig signers", err)
+				}
+				for _, s := range signers {
+					signerId := lookupId(s)
+					if m, found := accountsToMultisigs[signerId]; found { //add msig id to signer's collection
+						m = append(m, addr)
+						accountsToMultisigs[signerId] = m
+					} else {
+						n := []address.Address{addr}
+						accountsToMultisigs[signerId] = n
+					}
+				}
+
+				locked, err := ms.LockedBalance(abi.ChainEpoch(height))
+				if err != nil {
+					return xerrors.Errorf("failed to compute locked multisig balance: %w", err)
+				}
+
+				threshold, _ := ms.Threshold()
+				info := msigBriefInfo{
+					ID:        addr,
+					Signer:    signers,
+					Balance:   big.Min(big.Zero(), types.BigSub(act.Balance, locked)),
```

Comment:
pending https://filecoinproject.slack.com/archives/G013P3VU9CL/p1664331051760219

Reply:
Private slack, can't open, but it is almost certainly wrong...

Reply:
@ribasushi Can you elaborate here please? The Slack discussion is just about whether we want total or available balance here -- we settled on available.

Given that, is this wrong? Seems like the right way to get available balance to me?

Reply:
@arajasek I was flagging the outer part:

avail = Balance-locked ( ideally a positive number )

Then you take:
`min( zero, ideally_positive_number)`

which gives you... always zero or unlikely-negative ❓

Reply:
ah, thank you...i forgot that `min` and `max` aren't actually the same thing :D

Reply:
So just to confirm- we are taking spendable balance for now.